src/telemetry/telemetry_buffer.ts

```diff
@@ -0,0 +1,79 @@
+// Interfaces
+interface TelemetryEvent  {
+    timestamp : number,
+    action: TypeGenieTelemetryBuffer.Action,
+    value: any
+}
+
+interface TypeGenieTelemetryBufferInterface {
+
+    createTelemetryEvent(timestamp: number, action: TypeGenieTelemetryBuffer.Action, value: any) : TelemetryEvent;
+    addEvent(event: TelemetryEvent) : void;
+    removeEvent(index: number) : void;
+
+}
+
+
+
+export class TypeGenieTelemetryBuffer implements TypeGenieTelemetryBufferInterface{
+
+    private _events : TelemetryEvent[];
+
+    get events() {
+        return this._events;
+    }
+
+    set events(value : TelemetryEvent[]) {
+        this._events = value;
+    }
+
+    constructor() {
+        this.events = [];
+        console.log('Initializing TipeGenieTelemetryBuffer')
+    }
+
+    addEvent(event: TelemetryEvent): void {
+        this._events.push(event)
+    }
+
+    removeEvent(index: number): void {
+        this._events.slice(index,1)
+    }
+
+    clearEvents(): void {
+        this.events = [];
+    }
+
+    createTelemetryEvent(timestamp: number, action: TypeGenieTelemetryBuffer.Action, value: any): TelemetryEvent {
+
+        return  {
+            action: action,
+            timestamp: timestamp,
+            value: value
+            }
+        }
+
+     startTelemetryReport(interval: number) {
+        const context = this;
+        setInterval(() => {
+            if(context.events.length > 0) {
+                console.log('Will report: ', context.events);
+                context.clearEvents();
+            } else {
+                console.log('No events to report');
+            }
+        }, interval)
+     }
+
+}
+
+
+// Namespace for actions
+export namespace TypeGenieTelemetryBuffer {
+    export enum Action {
+        ACCEPT_FIRST_CHAR,
```

Comment:
I am wondering if need other action then `FETCH_COMPLETION`. The reason being if I know the `state` of the editor
say `{timestamp: 1, text: 'Hello ', completion: 'Renato', keystroke: ' '} and then `{timestamp: 2, text: 'Hello R', completion: 'enato`, keystroke: `R`}, then I can already know that it is a case of overtype. 

Alteratively, for 
`{timestamp: 1, text: 'Hello ', completion: 'Renato', keystroke: ' '} and then `{timestamp: 1, text: 'Hello Renato', completion: '', keystroke: `Tab`} is a case of `ACCEPT_COMPLETION`. 

So in my mind, the trickier bit is to set up the incremental `state` of the editor.

Reply:
Got it, I'll try to focus more on this perspective of state transition and reduce the complexity/importance of other events. Perhaps this will require a different approach than the one I presented in this PR, but it is definetly feasible